connectivity/tests/encryption.go

```diff
@@ -0,0 +1,178 @@
+	client := t.Context().RandomClientPod()
+
+	var server check.Pod
+	for _, pod := range t.Context().EchoPods() {
+		// Make sure that the server pod is on another node than client
+		if pod.Pod.Status.HostIP != client.Pod.Status.HostIP {
+			server = pod
+			break
+		}
+	}
+
+	bgCtx, bgCancel := context.WithCancel(ctx)
+
+	// clientHost is a pod running on the same node as the client pod, just in
+	// the host netns.
+	clientHost := t.Context().HostNetNSPodsByNode()[client.Pod.Spec.NodeName]
+
+	// Determine on which netdev iface to capture pkts. In the case of tunneling,
+	// we don't expect to see unencrypted pkts on a corresponding tunneling iface,
+	// so the choice is obvious. In the native routing mode, we run
+	// "ip route get $SERVER_POD_IP" from the client pod's node.
+	iface := ""
+	tunnelFeat, ok := t.Context().Feature(check.FeatureTunnel)
+	if ok && tunnelFeat.Enabled {
+		iface = "cilium_" + tunnelFeat.Mode // E.g. cilium_vxlan
+	} else {
+		cmd := []string{"/bin/sh", "-c",
+			fmt.Sprintf("ip -o r g %s from %s | grep -oP '(?<=dev )[^ ]+'",
+				server.Pod.Status.PodIP, client.Pod.Status.PodIP)}
+		t.Debugf("Running %s", strings.Join(cmd, " "))
+		dev, err := clientHost.K8sClient.ExecInPod(ctx, clientHost.Pod.Namespace,
+			clientHost.Pod.Name, "", cmd)
+		if err != nil {
+			t.Fatalf("Failed to get IP route: %s", err)
+		}
+		iface = strings.TrimRight(dev.String(), "\n\r")
+	}
+
+	t.Debugf("Detected %s iface for communication among client and server nodes",
+		iface)
+
+	bgStdout := &safeBuffer{}
+	bgStderr := &safeBuffer{}
+	bgExited := make(chan struct{})
+	// Start kubectl exec in bg (=goroutine)
+	go func() {
+		// Run tcpdump with -w instead of directly printing captured pkts. This
+		// is to avoid a race after sending ^C (triggered by bgCancel()) which
+		// might terminate the tcpdump process before it gets a chance to dump
+		// its captures.
+		cmd := []string{
+			"tcpdump", "-i", iface, "--immediate-mode", "-w", "/tmp/foo.pcap",
+			// Capture pod egress traffic.
+			// Unfortunately, we cannot use "host %s and host %s" filter here,
+			// as IPsec recirculates replies to the iface netdev, which would
+			// make tcpdump to capture the pkts (false positive).
+			fmt.Sprintf("src host %s and dst host %s", client.Pod.Status.PodIP, server.Pod.Status.PodIP),
+			// Only one pkt is enough, as we don't expect any unencrypted pkt
+			// to be captured
+			"-c", "1"}
+		t.Debugf("Running in bg: %s", strings.Join(cmd, " "))
+		err := clientHost.K8sClient.ExecInPodWithWriters(bgCtx, clientHost.Pod.Namespace,
+			clientHost.Pod.Name, "", cmd, bgStdout, bgStderr)
+		if err != nil {
+			t.Fatalf("Failed to execute tcpdump: %s", err)
+		}
+		close(bgExited)
+	}()
+
+	// Wait until tcpdump is ready to capture pkts
+	timeout := time.After(5 * time.Second)
+	for found := false; !found; {
+		select {
+		case <-timeout:
+			t.Fatalf("Failed to wait for tcpdump to be ready")
+		default:
+			line, err := bgStdout.ReadString('\n')
```

Comment:
Instead of concurrently reading from `bgStdout`, couldn't the bg goroutine check for tcpdump to be read and then just signal to this gorouting using a channel? This would probably allow to get rid of `safeBuffer` altogether. Or am I missing something?

Reply:
Good idea, but unfortunately the `ExecInPod*` blocks, and returns only after either a packet has been captured or ^C.